source/opt/block_merge_pass.cpp

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2017 The Khronos Group Inc.
+// Copyright (c) 2017 Valve Corporation
+// Copyright (c) 2017 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "block_merge_pass.h"
+
+#include "iterator.h"
+
+static const int kSpvEntryPointFunctionId = 1;
+
+namespace spvtools {
+namespace opt {
+
+bool BlockMergePass::IsLoopHeader(ir::BasicBlock* block_ptr) {
+  auto iItr = block_ptr->end();
+  --iItr;
+  if (iItr == block_ptr->begin())
+    return false;
+  --iItr;
+  return iItr->opcode() == SpvOpLoopMerge;
+}
+
+bool BlockMergePass::MergeBlocks(ir::Function* func) {
+  bool modified = false;
+  for (auto bi = func->begin(); bi != func->end(); ) {
+    // Do not merge loop header blocks, at least for now.
+    if (IsLoopHeader(&*bi)) {
```

Comment:
The comments could probably be clearer here. We are looking for a sequence of two blocks where the first has only one successor and the second has only one predecessor. Here, bi is the first block. The first block can have multiple preds and still be a candidate. And I think almost by definition a loop header can only have one successor. So theoretically, this case could occur, but it wasn't a priority for me at the moment and I didn't want to deal with extra logic of moving the merge op or checking that the next block wasn't a header. This logic was easier.

Also I don't think that this logic causes a requirement for structured control flow for this pass, it just doesn't optimize a particular structured construct at this time.